File: src/stores/connection.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import type { ConnectionConfig, ConnectionStatus, TestConnectionRequest } from '../types'
import { connectionApi } from '../api'

export const useConnectionStore = defineStore('connection', () => {
  const connections = ref<ConnectionConfig[]>([])
  const activeConnectionId = ref<string | null>(null)
  const connectionStatuses = ref<Map<string, ConnectionStatus>>(new Map())
  const loading = ref(false)
  const error = ref<string | null>(null)

  const activeConnection = computed(() => {
    if (!activeConnectionId.value) return null
    return connections.value.find(c => c.id === activeConnectionId.value) || null
  })

  const isConnected = computed(() => {
    if (!activeConnectionId.value) return false
    const status = connectionStatuses.value.get(activeConnectionId.value)
    return status?.connected || false
  })

  async function loadConnections() {
    loading.value = true
    error.value = null
    try {
      connections.value = await connectionApi.getConnections()
    } catch (err) {
      error.value = err instanceof Error ? err.message : '加载连接失败'
      console.error('Failed to load connections:', err)
    } finally {
      loading.value = false
    }
  }

  async function saveConnection(config: ConnectionConfig) {
    loading.value = true
    error.value = null
    try {
      const saved = await connectionApi.saveConnection(config)
      const index = connections.value.findIndex(c => c.id === saved.id)
      if (index >= 0) {
        connections.value[index] = saved
      } else {
        connections.value.push(saved)
      }
      return saved
    } catch (err) {
      error.value = err instanceof Error ? err.message : '保存连接失败'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function deleteConnection(id: string) {
    loading.value = true
    error.value = null
    try {
      await connectionApi.deleteConnection(id)
      connections.value = connections.value.filter(c => c.id !== id)
      if (activeConnectionId.value === id) {
        activeConnectionId.value = null
      }
    } catch (err) {
      error.value = err instanceof Error ? err.message : '删除连接失败'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function testConnection(request: TestConnectionRequest) {
    loading.value = true
    error.value = null
    try {
      const status = await connectionApi.testConnection(request)
      return status
    } catch (err) {
      error.value = err instanceof Error ? err.message : '测试连接失败'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function connect(connectionId: string) {
    loading.value = true
    error.value = null
    try {
      const status = await connectionApi.connect(connectionId)
      connectionStatuses.value.set(connectionId, status)
      if (status.connected) {
        activeConnectionId.value = connectionId
      }
      return status
    } catch (err) {
      error.value = err instanceof Error ? err.message : '连接失败'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function disconnect(connectionId: string) {
    loading.value = true
    error.value = null
    try {
      const status = await connectionApi.disconnect(connectionId)
      connectionStatuses.value.set(connectionId, status)
      if (activeConnectionId.value === connectionId) {
        activeConnectionId.value = null
      }
      return status
    } catch (err) {
      error.value = err instanceof Error ? err.message : '断开连接失败'
      throw err
    } finally {
      loading.value = false
    }
  }

  async function checkConnectionStatus(connectionId: string) {
    try {
      const status = await connectionApi.getConnectionStatus(connectionId)
      connectionStatuses.value.set(connectionId, status)
      return status
    } catch (err) {
      console.error('Failed to check connection status:', err)
      return null
    }
  }

  function setActiveConnection(connectionId: string | null) {
    activeConnectionId.value = connectionId
  }

  function clearError() {
    error.value = null
  }

  return {
    connections,
    activeConnectionId,
    connectionStatuses,
    loading,
    error,
    activeConnection,
    isConnected,
    loadConnections,
    saveConnection,
    deleteConnection,
    testConnection,
    connect,
    disconnect,
    checkConnectionStatus,
    setActiveConnection,
    clearError
  }
})

File: src-tauri/src/commands/connection.rs
use crate::db::pool::ConnectionPool;
use crate::error::Result;
use crate::models::connection::{ConnectionConfig, ConnectionStatus, TestConnectionRequest};
use crate::services::connection_manager::ConnectionManager;
use crate::services::postgres_service::PostgresService;

#[tauri::command]
pub async fn test_connection(request: TestConnectionRequest) -> Result<ConnectionStatus> {
    let config = ConnectionConfig {
        id: "test".to_string(),
        name: "test".to_string(),
        host: request.host,
        port: request.port,
        database: request.database,
        username: request.username,
        password: request.password,
        ssl_mode: request.ssl_mode,
        created_at: None,
        updated_at: None,
    };

    match ConnectionPool::test_connection(&config).await {
        Ok(_) => Ok(ConnectionStatus {
            id: "test".to_string(),
            connected: true,
            message: "Connection successful".to_string(),
            database_version: None,
        }),
        Err(e) => Ok(ConnectionStatus {
            id: "test".to_string(),
            connected: false,
            message: format!("Connection failed: {}", e),
            database_version: None,
        }),
    }
}

#[tauri::command]
pub async fn save_connection(config: ConnectionConfig) -> Result<ConnectionConfig> {
    ConnectionManager::save_connection(config)
}

#[tauri::command]
pub fn get_connections() -> Result<Vec<ConnectionConfig>> {
    ConnectionManager::get_all_connections()
}

#[tauri::command]
pub fn delete_connection(id: String) -> Result<()> {
    ConnectionManager::delete_connection(&id)
}

#[tauri::command]
pub async fn connect(connection_id: String) -> Result<ConnectionStatus> {
    let config = ConnectionManager::get_connection(&connection_id)?;
    
    match ConnectionPool::connect(connection_id.clone(), &config).await {
        Ok(_) => {
            let version = PostgresService::get_database_version(&connection_id).await.ok();
            Ok(ConnectionStatus {
                id: connection_id,
                connected: true,
                message: "Connected successfully".to_string(),
                database_version: version,
            })
        }
        Err(e) => Ok(ConnectionStatus {
            id: connection_id,
            connected: false,
            message: format!("Connection failed: {}", e),
            database_version: None,
        }),
    }
}

#[tauri::command]
pub fn disconnect(connection_id: String) -> Result<ConnectionStatus> {
    ConnectionPool::disconnect(&connection_id)?;
    Ok(ConnectionStatus {
        id: connection_id,
        connected: false,
        message: "Disconnected".to_string(),
        database_version: None,
    })
}

#[tauri::command]
pub fn get_connection_status(connection_id: String) -> Result<ConnectionStatus> {
    let connected = ConnectionPool::is_connected(&connection_id);
    Ok(ConnectionStatus {
        id: connection_id,
        connected,
        message: if connected {
            "Connected".to_string()
        } else {
            "Not connected".to_string()
        },
        database_version: None,
    })
}

#[tauri::command]
pub fn get_active_connections() -> Result<Vec<String>> {
    Ok(ConnectionPool::get_active_connections())
}
